### DISCUSSION TITLE
Possible to work with JK-PB2A16S20P Inverter BMS?

### DISCUSSION THREAD

**Discussion by virus100b** [Sleeper85/esphome-jk-bms-can#52] [General]
Hello,

As title say, i am wondering if this project can work with JK-PB2A16S20P new type inverter BMS from JK?

**Comment by Sleeper85** [comment]
Hi,

Yes with the new version in the development branch. This is already in use with someone who owns 7x JK-BMS-PB.

**Comment by virus100b** [comment]
Good News! I am planning to  change my BMS to PB series and i don't want to loose HA communication.
Where i should connect ESP on 485 port? On the port for inverter or for the BMS parallel port at master BMS?

**Reply by Sleeper85** [reply]
This is a draft, this is what was achieved during the first test with an Atom S3 Lite.

[JK-BMS-PB-CAN solution](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/documents/README/JK-BMS-PB-CAN_solution.md)

**Reply by virus100b** [reply]
Thank You for info. I have read the docs.  It looks promising!

**Comment by virus100b** [comment]
Compiling PB mode 2  version i get this error. I have copied packages folder in the Esphome folder. What i am doing wrong? 

Compiling .pioenvs/smartbms/src/main.o
In file included from src/esphome/components/jk_rs485_sniffer/jk_rs485_sniffer.h:6,
                 from src/esphome/components/jk_rs485_bms/jk_rs485_bms.h:8,
                 from src/esphome.h:32,
                 from src/main.cpp:3:
src/esphome/components/jk_rs485_sniffer/binary_output.h:18:7: error: redefinition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
In file included from src/esphome/components/binary/light/binary_light_output.h:4,
                 from src/esphome.h:15,
                 from src/main.cpp:3:
src/esphome/components/output/binary_output.h:18:7: note: previous definition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
Generating LD script .pioenvs/smartbms/memory.ld
Compiling .pioenvs/smartbms/app_trace/app_trace.o
*** [.pioenvs/smartbms/src/main.o] Error 1
========================= [FAILED] Took 607.51 seconds =========================

**Reply by Sleeper85** [reply]
This example file works by default with Atom S3 and 7 bms.
Have you correctly configured your YAML for your board and your number of BMS?

**Reply by virus100b** [reply]
> This example file works by default with Atom S3 and 7 bms.

Indeed it compile ok for S3 but i am not heaving a S3 board. I have C3 and, ESP WROOM 32  and for those i get errors compiling. I have created a new board file and configured for C3 board and pins but it failed to compile successfully with same error sa for ESP WOOM 32.

@Sleeper85 When you have time please check the file for ESP Wroom 32 and if i am not asking to much maybe you can create one for C3 also. Thank you in advance!

**Comment by virus100b** [comment]
Yes. I have connfigured for esp32 

packages:
  # Board : uncomment your board
  device_board: !include packages/board/board_esp32-devkit-v1.yaml
  # device_board: !include packages/board/board_esp32-s3-devkitc-1.yaml
  #device_board: !include packages/board/board_atom-s3-lite.yaml

and   sniffer1: !include
    file: packages/bms/bms_jk-bms_rs485_pb_sniffer.yaml
    vars:
      sniffer_id: 'sniffer1'
      sniffer_protocol: 'JK02_32S'
      sniffer_uart_id: 'uart_esp_1'
      sniffer_talk_pin: 4 # ESP32: 4, ESP32-S3: 8

I will have just 2 BMS + ESP as master but i did not exclude the rest of  BMS-es

**Comment by virus100b** [comment]
So, I delete BMS 3 to 7, left just BMS 1 & 2.  Now i get the fallowing error:

INFO ESPHome 2024.3.0
INFO Reading configuration /config/esphome/jk-pb.yaml...
INFO Updating https://github.com/txubelaxu/esphome-jk-bms.git@main
WARNING GPIO2 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Detected timezone 'Europe/Bucharest'
INFO Generating C++ source...
INFO Compiling app...
Processing smartbms (board: esp32doit-devkit-v1; framework: espidf; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40406.240122 (4.4.6) 
 - tool-cmake @ 3.16.4 
 - tool-ninja @ 1.7.1 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Reading CMake configuration...
No dependencies
Compiling .pioenvs/smartbms/src/main.o
Archiving .pioenvs/smartbms/esp-idf/asio/libasio.a
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_common.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_common_loader.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_clock_init.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_flash.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_mem.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_random.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_random_esp32.o
In file included from src/esphome/components/jk_rs485_sniffer/jk_rs485_sniffer.h:6,
                 from src/esphome/components/jk_rs485_bms/jk_rs485_bms.h:8,
                 from src/esphome.h:32,
                 from src/main.cpp:3:
src/esphome/components/jk_rs485_sniffer/binary_output.h:18:7: error: redefinition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
In file included from src/esphome/components/binary/light/binary_light_output.h:4,
                 from src/esphome.h:15,
                 from src/main.cpp:3:
src/esphome/components/output/binary_output.h:18:7: note: previous definition of 'class esphome::output::BinaryOutput'
 class BinaryOutput {
       ^~~~~~~~~~~~
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_utility.o
Compiling .pioenvs/smartbms/bootloader_support/src/esp_image_format.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_encrypt.o
Compiling .pioenvs/smartbms/bootloader_support/src/secure_boot.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_partitions.o
Compiling .pioenvs/smartbms/bootloader_support/src/flash_qio_mode.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_flash_config_esp32.o
Compiling .pioenvs/smartbms/bootloader_support/src/bootloader_efuse.o
Compiling .pioenvs/smartbms/bootloader_support/src/idf/bootloader_sha.o
Archiving .pioenvs/smartbms/esp-idf/bootloader_support/libbootloader_support.a
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborencoder_close_container_checked.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborencoder.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborerrorstrings.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborparser_dup_string.o
Compiling .pioenvs/smartbms/cbor/tinycbor/src/cborparser.o
*** [.pioenvs/smartbms/src/main.o] Error 1
========================= [FAILED] Took 55.50 seconds =========================

**Reply by Sleeper85** [reply]
Clean esphome build files.

Don't forget to set a talk_pin, hope you have the right RS485 converter.

Don't forget to configure the dip switches of your BMS.

- mode1: 0x00 and 0x01
- mode2: 0x01 and 0x02

**Comment by virus100b** [comment]
I have clean build files, i have set the talk pin as :      sniffer_talk_pin: 4 # ESP32: 4, ESP32-S3: 8
I have the proper RS485 converter but i cannot use as the esp cannot be flashed, still get the error while compile.

**Reply by Sleeper85** [reply]
Hi,

I responded to your message on the forum. I also created an ESP32-C3 board and defined the gpios to use.

https://diysolarforum.com/threads/jk-bms-can-with-new-cut-off-charging-logic-open-source.79325/page-56#post-1109394

https://diysolarforum.com/threads/jk-bms-can-with-new-cut-off-charging-logic-open-source.79325/page-56#post-1109421

**Comment by Sleeper85** [comment]
Indeed I encounter the same compilation problem with ESP32 (talk pin 4) however I was able to compile the example file `multi-bms_RS485_JK-BMS_PB_series_mode2_ESP32-C3.yaml` without problems for ESP32-C3 (talk pin 10).

**Reply by virus100b** [reply]
I have compiled successfully the file board_esp32-s3-devkitc-1 modifing the pins for C3 bord. Looks like there is a problem in the lights block but changing this part:
light:
  # ESP Light
  - platform: binary
    output: esp_led
    id: esp_light
    name: "${name} On board light"
    internal: true

with this one manage to compile.

light:
  # https://esphome.io/components/light/esp32_rmt_led_strip
  - platform: esp32_rmt_led_strip
    id: esp_light
    name: ${name} ESP Light
    rgb_order: GRB
    pin: 48
    num_leds: 1
    rmt_channel: 0
    chipset: ws2812
    entity_category: config
    internal: true

Currently i have only one JK BMS  set as master. I can see most of entities am am interested (some appear as unknow), but i have issues with switches. If i am turning on and off from BMS bluetooth app i can see in Home Assistant the change, but when i am turning off or on from HA they are not changing in BMS app. Is this change one way (just from JK Bluetooth App to HA), if not what is the fix for that? 

I am open to test any files updates. Thanks!

**Comment by holle75** [comment]
Also very interested in the solution for the JK-PB*.
Also which (different) hardware might be needed to go wired and i have no idea if my CAN-Bus is 5V or 3.3V.
At the moment i use syssi via BT just for documentation/info in fhem. This works nicely. When starting to modify CAN-Information (Studer Inverter with the wrong SOC100/stop charging/no balancing problem) to the inverter i would prefer a wired solution (without loosing info to fhem).

**Comment by Sleeper85** [comment]
@holle75 

[YamBMS README](https://github.com/Sleeper85/esphome-jk-bms-can/tree/development)

[JK-PB RS485 solution](https://github.com/Sleeper85/esphome-jk-bms-can/blob/development/documents/README/BMS_JK-PB_RS485_solution.md)
